src/library.js

```diff
@@ -41,10 +41,13 @@ LibraryManager.library = {
 
   $stringToNewUTF8__deps: ['malloc'],
   $stringToNewUTF8: function(jsString) {
-    var length = lengthBytesUTF8(jsString)+1;
-    var cString = _malloc(length);
-    stringToUTF8(jsString, cString, length);
-    return cString;
+    if (jsString) {
```

Comment:
Is this intended to handle an empty string, a null, or both? Atm it will handle both, but it seems odd to turn an empty string into a 0 pointer?

Reply:
The intent is to handle both, I think it is a good optimization to avoid running dummy `malloc(0)`s on empty strings? That is something that one would generally do in C programs to optimize empty strings as well.

Reply:
Sounds reasonable. But it means we need to audit all the callers of this method to verify that they handle a 0 now?

Reply:
I did go through all the call sites of the function, and the test suite seems to agree. After all, JS side code does not really have any use of the returned integer value, its only use will be to flow that number directly to wasm (maybe cache it in some structure).

Reply:
Makes sense. What about all the wasm users, though? Calling `free(0)` is fine, but if they do something like `strcpy` or `printf` of NULL that could error.

Reply:
In musl a printf of null string ptr prints out `(null)`. strcpy of null would be a problem.

stringToNewUTF8() is not publicly documented out to external users (though that would not prevent it from being used externally). If external users unconditionally strcpying is worrying, we need to reconsider. In such case, maybe `stringToNewUTF8(null/undefined)` should also allocate a dummy pointer that holds a zero.

Reply:
The same would be needed for `stringToNewUTF8('')` (the empty string)? If so then that would revert the change to this file.

Overall this seems a little risky to me for little benefit (as it increases code size in this file acually). Is it worth it for the benefits in `library_webgl.js` to figure out a safe way to avoid allocating for falsey inputs here?

Reply:
> Is it worth it for the benefits in `library_webgl.js` to figure out a safe way to avoid allocating for falsey inputs here?

The reason for placing the check inside `stringToNewUTF8()` is the same as why [UTF8ToString() checks for null input](https://github.com/emscripten-core/emscripten/blob/cbc974264e0b0b3f0ce8020fb2f1861376c66545/src/runtime_strings.js#L112): it is better for code size to have that check exist only once for all callers, rather than require all callers to pre-validate first.

Updated the code to always return an allocated string.

> Overall this seems a little risky to me for little benefit (as it increases code size in this file acually).

This is the right location to fix this. Otherwise we would either increase code size for all callers when the would each need to validate, or require that callers statically write code that never produce empty strings. Only if we know that our users build programs that do not have any call sites that feed in falsey strings, it would be smaller for general code size. If they had one, it would be all the same whether `stringToNewUTF8()` checks or if caller checks, and if they have two or more, it is more optimal to check here. So we have to pick for the types of programs we optimize size for: are we optimizing size for programs that have 0, 1 or >= 2 such call sites?

Reply:
Actually, maybe we don't have any other callers that do need to rely on `stringToNewUTF8(null/undefined/false/0)` to work. Hoisted the check to library_webgl.js on that assumption. If we do get more of such call sites, then it would be best to locate the check inside `stringToNewUTF8()`.